apps/backend/src/controllers/agri/bpp/cancel.ts

```diff
@@ -0,0 +1,125 @@
+import { NextFunction, Request, Response } from "express";
+import {
+	responseBuilder,
+	redisFetchFromServer,
+	send_nack,
+	updateFulfillments,
+} from "../../../lib/utils";
+import { ERROR_MESSAGES } from "../../../lib/utils/responseMessages";
+import { ON_ACTION_KEY } from "../../../lib/utils/actionOnActionKeys";
+import {
+	ORDER_CACELLED_BY,
+	ORDER_STATUS,
+} from "../../../lib/utils/apiConstants";
+
+export const cancelController = async (
+	req: Request,
+	res: Response,
+	next: NextFunction
+) => {
+	try {
+		const { scenario } = req.query;
+		const { transaction_id } = req.body.context;
+		const on_confirm_data = await redisFetchFromServer(
+			ON_ACTION_KEY.ON_CONFIRM,
+			transaction_id
+		);
+		if (!on_confirm_data) {
+			return send_nack(res, ERROR_MESSAGES.ON_CONFIRM_DOES_NOT_EXISTED);
+		}
+		if (on_confirm_data.message.order.id != req.body.message.order_id) {
+			return send_nack(res, ERROR_MESSAGES.ORDER_ID_DOES_NOT_EXISTED);
+		}
+
+		const on_search_data = await redisFetchFromServer(
+			ON_ACTION_KEY.ON_SEARCH,
+			transaction_id
+		);
+
+		const item_measure_ids =
+			on_search_data.message.catalog['bpp/providers'][0].items.reduce(
```

Comment:
This should not be picked from on_search. Use on_confirm or any later actions here instead.